### ISSUE TITLE
names client Resolver class not handling connectionLost

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4015]
|<img alt="pythonologist's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| pythonologist reported|
|-|-|
|Trac ID|trac#4015|
|Type|defect|
|Created|2009-09-11 07:28:35Z|

In some circumstances, a twisted.names.client.Resolver instance has the non-existent method connectionLost invoked. Here's a traceback:

```
Traceback (most recent call last):
  File "xmlharness.py", line 724, in run_program
    reactor.run()
  File "C:\software\python26\lib\site-packages\twisted-8.2.0\twisted\internet\base.py", line 1128, in run
    self.mainLoop()
  File "C:\software\python26\lib\site-packages\twisted-8.2.0\twisted\internet\base.py", line 1140, in mainLoop
    self.doIteration(t)
  File "C:\software\python26\lib\site-packages\twisted-8.2.0\twisted\internet\selectreactor.py", line 140, in doSelect
    _logrun(selectable, _drdw, selectable, method, dict)
--- <exception caught here> ---
  File "C:\software\python26\lib\site-packages\twisted-8.2.0\twisted\python\log.py", line 84, in callWithLogger
    return callWithContext({"system": lp}, func, *args, **kw)
  File "C:\software\python26\lib\site-packages\twisted-8.2.0\twisted\python\log.py", line 69, in callWithContext
    return context.call({ILogContext: newCtx}, func, *args, **kw)
  File "C:\software\python26\lib\site-packages\twisted-8.2.0\twisted\python\context.py", line 59, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "C:\software\python26\lib\site-packages\twisted-8.2.0\twisted\python\context.py", line 37, in callWithContext
    return func(*args,**kw)
  File "C:\software\python26\lib\site-packages\twisted-8.2.0\twisted\internet\selectreactor.py", line 156, in _doReadOrWrite
    self._disconnectSelectable(selectable, why, method=="doRead")
  File "C:\software\python26\lib\site-packages\twisted-8.2.0\twisted\internet\posixbase.py", line 190, in _disconnectSelectable
    selectable.readConnectionLost(f)
  File "C:\software\python26\lib\site-packages\twisted-8.2.0\twisted\internet\tcp.py", line 509, in readConnectionLost
    self.connectionLost(reason)
  File "C:\software\python26\lib\site-packages\twisted-8.2.0\twisted\internet\tcp.py", line 680, in connectionLost
    Connection.connectionLost(self, reason)
  File "C:\software\python26\lib\site-packages\twisted-8.2.0\twisted\internet\tcp.py", line 520, in connectionLost
    protocol.connectionLost(reason)
  File "C:\software\python26\lib\site-packages\twisted-8.2.0\twisted\names\dns.py", line 1769, in connectionLost
    self.controller.connectionLost(self)
exceptions.AttributeError: Resolver instance has no attribute 'connectionLost'

```

I haven't got a test program as such, just iterate over a variety of domain names. Possibly the DNS server is dropping the connection prior to responding?

Attachments:

* [twisted.names.client.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e5c/e5ceae329a2fab85c2a598dd538b4e73f6f98f1f/e0bf710b928ad913e0bc4ce0bc8af92e804855ae.patch) (419 bytes) - added by Alrond on 2009-12-03 14:02:28Z - 
* [twistd_log.txt](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e5c/e5ceae329a2fab85c2a598dd538b4e73f6f98f1f/de9b7af95cc8a35a509353b9023f1dda22fe715b.txt) (5089 bytes) - added by teratorn on 2011-10-20 14:46:00Z - 
* [dig_log.txt](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e5c/e5ceae329a2fab85c2a598dd538b4e73f6f98f1f/de709c7722f27582a5287a830414fbd388288cca.txt) (2550 bytes) - added by teratorn on 2011-10-20 14:46:12Z - 
* [dig@10.0.0.1.pcap](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e5c/e5ceae329a2fab85c2a598dd538b4e73f6f98f1f/65fdb3c28e506a8eb20ccb827d841eaf3820e2e9.pcap) (1364 bytes) - added by teratorn on 2011-10-20 14:46:25Z - 
* [t.names_to_71.42.153.1.pcap](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e5c/e5ceae329a2fab85c2a598dd538b4e73f6f98f1f/82b8850cabe172de03ffe6fa658ed83915f21f6f.pcap) (2894 bytes) - added by teratorn on 2011-10-20 14:46:34Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4015 4015
type__defect defect
reporter__pythonologist pythonologist
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__duplicate duplicate
component__names names
keywords__ 
time__1252654115000000 1252654115000000
changetime__1347194777000000 1347194777000000
version__None None
owner__ 
cc__teratorn@... cc__rwall
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="Alrond's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|Alrond set status to `closed`|
|-|-|

I confirm this problem with the same error output.
Temporary may be solved with patch:

--- client.py_original  2009-12-03 15:01:03.000000000 +0100
+++ client.py   2009-12-03 15:01:27.000000000 +0100
@@ -243,6 +243,8 @@
             self.queryTCP(q, t).chainDeferred(d)
         del self.pending[:]

+    def connectionLost(self, protocol):
+        pass

     def messageReceived(self, message, protocol, address = None):
         log.msg("Unexpected message (%d) received from %r" % (message.id, address))

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun set status to `reopened`|
|-|-|

The problem actually has to be fixed before the ticket can be closed.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="teratorn's avatar" src="https://avatars.githubusercontent.com/u/154069?s=50" width="50" height="50">](https://github.com/teratorn)<a name="note_4"></a>|@teratorn commented|
|-|-|

I'm hitting this issue, and it seems to be the cause or a symptom of a bug which prevents recursive name lookups from working through a t.names DNS server.

I'm attaching trace files that clearly show the failure. The test scenario is [LAN PC] <--> [GATEWAY PC RUNNING T.NAMES] <--> [ISP DNS SERVER]

LAN PC is 10.0.0.222; GATEWAY is 10.0.0.1 (internal) and 71.42.153.210 (external); ISP DNS is 71.42.153.1

Using dig, an A record lookup is performed for www.google-analytics.com, which succeeds. Several minutes are allowed to pass, and then the exception appears in the twistd log file. Then a second A record lookup is performed for www.google-analytics.com, which fails.

**Comment by twisted-trac** [comment]
|<img alt="buddyns's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|buddyns commented|
|-|-|

The exception here is raised by twisted's Protocol class to the DNSProtocol object used by Names for TCP queries. DNSProtocol deliberately propagates to its controller any connectionLost() call.

The fix of adding a dummy connectionLost() method to Resolver works. An implementation where no empty method is called is more flexible. In names/dns.py :

```
def connectionLost(self, reason):
    self.controller.connectionLost(self)
```

... to become:

```
def connectionLost(self, reason):
    if callable(getattr(self.controller, 'connectionLost')):
        self.controller.connectionLost(self)
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

[#5224](https://github.com/twisted/twisted/issues/5224) was a duplicate of this.

**Comment by twisted-trac** [comment]
|<img alt="buddyns's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|buddyns commented|
|-|-|

We have been using this solution successfully for a few weeks:

in `names/dns.py`:

```
def connectionLost(self, reason):
    """
    Notify the controller that this protocol is no longer
    connected.
    """
    if callable(getattr(self.controller, 'connectionLost', None)):
        self.controller.connectionLost(self)
```

If there's no arguments against, please merge and let's close this ticket.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_8"></a>|@wallrj set status to `closed`|
|-|-|

Looks like this has already been resolved in [#5471](https://github.com/twisted/twisted/issues/5471) (r33527)

Closing as a duplicate.
